### ISSUE TITLE
Option to align route params with schema at startup

### ISSUE THREAD

**Issue by avele** [fastify/fastify#5215]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

Introduce startup validation in Fastify to check consistency between route parameters in path strings and schema definitions.

### Motivation

Ensuring route parameter consistency to prevent runtime errors

### Example

```js
// Inconsistent route params and schema could lead to unnoticed issues until runtime
fastify.get('/item/:itemId', {
  schema: {
    params: {
      type: 'object',
      properties: {
        id: { type: 'string' } // Potential issue: 'itemId' in route vs 'id' in schema
      }
    }
  },
  handler: (request, reply) => { /* ... */ }
});
```

A pre-runtime check could catch these mismatches early, enhancing developer experience and application stability.

**Comment by metcoder95** [comment]
👋 
I wouldn't add this to the core as it can make users make wrong assumptions and noise while working with the framework and setting its routes.

Rather this sounds like a good idea for a plugin, would you like to create one?

**Comment by Eomm** [comment]
We have something similar here:
- https://github.com/fastify/fastify-swagger/pull/761
But it does not throw if there is a mismatch.

The hardest thing is to list the path parameters because they are a router's internal info

**Comment by sf3ris** [comment]
👋 Hi,
Was taking a look at this issue since I had the same problem a few times and it has been quite time-consuming to debug.
Since the parameters are router's internal info, do you think could be feasible to validate right before adding the route within the `route.addNewRoute` function? Maybe passing a flag if we dont want to be too strict?

I guess to make a standalone plugin we need somehow to leak the router's internal or expose them.
What you think? 🤔

**Comment by metcoder95** [comment]
I'd say that leaking the router as is can be pretty dangerous as it's easy to alter it and make `fastify` fail, causing issues to spawn within the project (unless we assess is not a problem).

Potentially, have we already considered exposing an abstraction of `fmw.findRoute`?

I can see potential use cases there (like this plugin). Although I've not tested if it returns the `parameters` of the route path as well.

**Comment by sf3ris** [comment]
Makes sense, Ill try to give it a shot to see whether we can abstract it so that other plugins can use it.
So, just to recap, do you think that the reported case above should not throw on the fastify core without any plugin?
The plugin should only check that the parameters declared on the schema should match the ones in the URL?

**Comment by sf3ris** [comment]
Tried a quick implementation for opening a discussion wether is better to extract into a standalone project (what scope would it add? just to validate routes with params?)
see: https://github.com/fastify/fastify/pull/5230

**Comment by metcoder95** [comment]
The PR LGTM overall, but I'm over the fence with having the validation within core. This seems to have pretty specific use cases, and the harm of unexpected behaviour can cause friction.

That's one of the main reasons to suggest develop a community plugin instead

**Comment by sf3ris** [comment]
Should this be closed due to https://github.com/fastify/fastify/pull/5230 or do we want to keep it open for the plugin discussion?

**Comment by Uzlopak** [comment]
I think it should be in core. It feels somehow wrong to load a plugin, to "just" check if routes are registered correctly.

**Comment by mcollina** [comment]
We support multiple validators, and we can't safely implement this.

**Comment by metcoder95** [comment]
> Should this be closed due to #5230 or do we want to keep it open for the plugin discussion?

I believe we should keep it open for plugin discussion

**Comment by sf3ris** [comment]
To me, a nice plugin I would use to validate the parameters would be:

### Plugin proposal adding a new hook
- Fastify core could add a new event like `onRegister` emitted every time a new route is registered (If its already something similar, better)
- The plugin, after being registered, will automatically add a new hook `onRegister` that will validate the route parameters.

#### Advantages
- The `onRegister` hook can be used by any other plugin that wants to validate or performs any action after every route is registered (logging, etc ..)
- The user of the plugin doesn't need to validate each route every time or calling like a `plugin.validateAll` method that could have lots of responsibility like:
    - how to retrieve all the routes
    - how to validate all the routes
 
 #### Cons
 - One more hook to add on the fastify core
 
 ### Plugin proposal using existent hooks
- The plugin, after being registered, will automatically add a new hook `preHandler` that will validate the route parameters before executing the route.
 E.g. if I call GET `/artists/123` during the preHandler the plugin will check if the `/artists/:id` will match a `:id` parameter inside the schema

#### Advantages
- No need to implement anything on the core
- Easy enough to use and the error will be meaningful for the user
- Only the not valid routes will throw while other ones will still keep working
 
 #### Cons
 - The route will be validated only once called, not right after registration.
 - The server will boot even if invalid routes

**Comment by Uzlopak** [comment]
We have already onRoute hook, which is triggered when a route is added.

**Comment by sf3ris** [comment]
Ouch, you're right. Thanks for the quick reply and the insight.
Then I think option one using the `onRoute` hook would fit the abovementioned perfectly.

What you think giving it a shot? Do you have any other community plugins that are responsible for validating routes just check them out?

**Comment by sf3ris** [comment]
@Uzlopak I tried to create the abovementioned plugin but I've encountered an issue, the `onRoute` hook is called before the actual registration.
In fact, doing the following steps will return null inside the hook.
```js
fastify.addHook('onRoute', (routeOpts) => {
    const {method, url} = routeOpts.
    const route = fastify.findRoute({method, url});
    console.log(route); // null
})

fastify.get('/index', (req, reply) => reply.send("OK"));
```

Any specific reason why the hooks are called [here](https://github.com/fastify/fastify/blob/2a6ee491a855117f1d48889232ab8d72607bbb3f/lib/route.js#L276) and the actual registration happens [here](https://github.com/fastify/fastify/blob/2a6ee491a855117f1d48889232ab8d72607bbb3f/lib/route.js#L349) ?

To overcome this without touching the core, I could parse the `routeOptions.url` and matching against `routeOptions.schema.params` but the fmw route object returned from the `fastify.findRoute` is more reliable and already has the parsed params on the object.

**Comment by Uzlopak** [comment]
I think it would be better to extract the parameter determination functionality in `find-my-way` 

https://github.com/delvedor/find-my-way/blob/a4d92627b818b42cb682fb2c808e1ca694b215f5/index.js#L171

and then use it.

But I wonder what you can do with regex values for route?!

**Comment by sf3ris** [comment]
> But I wonder what you can do with regex values for route?!
The reported issue here was that there are no checks if the registered params inside the route is the same as the one inside the declared schema.
E.g.
```js
fastify.get('/artists/:artistId', {
   schema: { params: { id: { type: 'integer' } } }
}, handler);
```

In this case the application would start correctly but hiding silent errors due to mismatch between declared params inside URL and the one declared in the schema.

The plugin to me, should check that the parsed params are the same as the ones declared inside the schema.
Does it make sense to you?

> I think it would be better to extract the parameter determination functionality in find-my-way

You mean exposing the functionality from fmw that will be exposed from fastify core (to avoid dependency leaks)?

**Comment by metcoder95** [comment]
Hmm, I think we are going down the rabbit hole; setting fmw aside to make the plugin fit the use case will be slightly more problematic than we thought.

Have you considered doing the registration by yourself using `fwm`?
i.e. the `onRoute` hook intercepts the registering, registers the route against a custom instance of `fmw`, obtains the route definition, does the validation and throws if mismatches.

You can also use `onReady` to throw away the `fwm` instance to avoid having dangling objects so it can be gc'ed.

**Comment by mcollina** [comment]
> Any specific reason why the hooks are called here and the actual registration happens here ?

Yes.

---

You don't need to call fmw. You have all the schema inside routeOptions.

**Comment by sf3ris** [comment]
> You don't need to call fmw. You have all the schema inside routeOptions.

Yes, you have the schema but you don't have the parsed params from the route path. 
For example, the fmw route object has the following fields that make the checks easier
```js
{ 
    schema: { params: { id: 'integer' } },
    params: { id: ':id' },
   ...otherFmwRouteFields
}
```

> Hmm, I think we are going down the rabbit hole; setting fmw aside to make the plugin fit the use case will be slightly more problematic than we thought.

I totally agree with this and I think is getting much more complicated than the problem it tries to solve.
Just wanted to give it a shot to close the loop of the reported issue but for me is totally fine to keep it as it is.

**Comment by climba03003** [comment]
Wouldn't it as simple as the following check?

```js
fastify.addHook('onRoute', function(option) {
  // check only when schema provided
  if(option.schema && option.schema.params) {
    for(const param of Object.keys(option.schema.params)) {
      // here is simple, named param actually always start with colon and follow the param name
      // so, what we need to do is check by includes
      if(!option.path.includes(`:${param}`)) throw Error(`${option.path} does not include param ${param}`)
    }
  }
})
```

**Comment by sf3ris** [comment]
Yes, this was exactly what we proposed along the thread but I fear that there will be a dependency leak between our param extraction and the one used by fmw internally.
Doing the parsing by ourselves will work and "probably" we would never introduce breaking changes, but a change/fix on the fmw parameter parsing (the real one then used to retrieve the value from a named parameter) could break projects using this validator.

To me, to create a stable plugin, the params parsed from the fmw should be exactly the same that we use to check against the schema.
Do you think could be an option to:
- add the fmw param extraction to the fmw interface
- when creating the `routeOptions` for the `onRoute` event, adding the parsed params to the object?
   e.g. 
   ```js
        const routeOptions = {
            ...routeOptions,
            params: fmw.extractParams(path)
        }
   ```

But, if you think could be worth to implement another extraction algorithm I can try to write a plugin to solve the issue 🙌

**Comment by climba03003** [comment]
@mcollina 
I though about it a while, but https://github.com/fastify/fastify/pull/5230 give too much power to the users.
It provide the whole route information including even the context within `fastify`. 
It also means that the read-only of context will be useless after this PR.

We need to re-considerate should be we expose all the information on route and hidden field on `fastify`.

**Comment by mahekfatima585-design** [comment]
Hi
I've created a community plugin for this issue.

Repository: https://github.com/mahekfatima585-design/fastify-route-params-validator

It validates route path parameters against schema.params at startup using onRoute + onReady hooks.

**Comment by LeonxLJX** [comment]
I'd like to take this one (`Option to align route params with schema at startup`). I'll look into the root cause and follow up with a PR. (claiming via @LeonxLJX)

**Comment by LeonxLJX** [comment]
I'd like to take this one (`Option to align route params with schema at startup`). I'll look into the root cause and follow up with a PR. (claiming via @LeonxLJX)
